src/components/date-picker/DatePicker.tsx

```diff
@@ -367,28 +434,72 @@ class DatePicker extends React.Component<DatePickerProps> {
         let inputDate: Date | null | undefined = null;
 
         if (this.dateInputEl) {
-            inputDate = new Date(this.dateInputEl.value);
+            let dateInputElVal = null;
+            if (isAccessible && !this.canUseDateInputType) {
+                dateInputElVal = this.parseDisplayDateType(this.dateInputEl.value);
+            }
+            inputDate = new Date(dateInputElVal || this.dateInputEl.value);
         }
 
-        if (isTextInputAllowed && inputDate && inputDate.getDate()) {
+        if ((isTextInputAllowed || (isAccessible && !this.canUseDateInputType)) && inputDate && inputDate.getDate()) {
             this.onSelectHandler(inputDate);
         }
     };
 
     handleButtonClick = (event: React.SyntheticEvent<HTMLButtonElement>) => {
         event.preventDefault();
         event.stopPropagation();
+        const { isAccessible, isDisabled } = this.props;
+
+        if (isAccessible) {
+            if (isDisabled || !this.datePicker) {
+                return;
+            }
+
+            if (this.datePicker.isVisible()) {
+                this.datePicker.hide();
+            } else {
+                this.datePicker.show();
+            }
+            return;
+        }
 
         if (!this.shouldStayClosed) {
             this.focusDatePicker();
         }
     };
 
+    handleOnClick = (event: React.SyntheticEvent<HTMLInputElement>) => {
+        const { isAccessible } = this.props;
+
+        if (isAccessible) {
+            // Suppress Firefox default behavior: clicking on input type "date"
+            // opens the browser date picker.
+            event.preventDefault();
+            event.stopPropagation();
+        }
+    };
+
     formatDisplay = (date?: Date | null): string => {
         const { displayFormat, intl } = this.props;
         return date ? intl.formatDate(date, displayFormat) : '';
     };
 
+    formatDisplayDateType = (date?: Date | null): string => {
+        // Input type "date" only accepts the format YYYY-MM-DD
+        return date ? date.toISOString().slice(0, 10) : '';
+    };
+
+    parseDisplayDateType = (dateString?: string | null): Date | null => {
+        if (dateString) {
+            // Calling new Date('YYYY-MM-DD') without 'T00:00:00' yields undesired results:
```

Comment:
Do we need the timezone part of the ISO string too?

Reply:
The timezone part is not required. It's only if the string is "date-only" that the issue occurs.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/Date#parameters (scroll to dateString)